autograd.cpp

```diff
@@ -386,6 +386,39 @@ void SGD::init_state() {
   momentum_buffers_.clear();
 }
 
+/// Adapted from
+/// https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py
+void Adagrad::step() {
+  for (auto& pair : model_->parameters()) {
+    auto& name = pair.first;
+    auto& grad = pair.second.grad();
+    auto& p = pair.second.data();
+    if (!grad.defined()) continue;
+
+    auto d_p = grad.data();
+    d_p.add_(p, weight_decay_);
```

Comment:
Wrap this in an if statement to not launch a CUDA kernel if weight_decay_ is 0.

Reply:
Ok, just copied from SGD.